### ISSUE TITLE
bug with MatrixSymbol and subs()

### ISSUE THREAD

**Issue by natethern** [sympy/sympy#11821]
This session illustrates the bug:

In [1]: import sympy
In [2]: sympy.__version__
Out[2]: '1.0'
In [3]: from sympy import *
In [4]: A=MatrixSymbol('A',1,3)
In [5]: B=MatrixSymbol('B',1,3)
In [6]: C=MatrixSymbol('C',1,3)
In [7]: E=A-B
In [8]: F=C[0,0]
In [9]: F.subs(C,E)
Out[9]: (-1)*B + A[0, 0]

The output should be 
Out[9]: (-1)*B[0, 0] + A[0, 0]
or
Out[9]: ((-1)*B + A)[0, 0]

**Comment by natethern** [comment]
Continuing the previous session:
In [10]: Fs=F.subs(C,E)
In [11]: Fs.func
Out[11]: sympy.matrices.expressions.matexpr.MatrixElement
In [12]: Fs.args
Out[12]: ((-1)*B + A, 0, 0)

It's apparent that the actual substitution is performed correctly and this is a bug in the printer.

**Comment by Upabjojr** [comment]
Hi @natethern , can you fix this?

Basically, it's about finding the `_print_MatrixElement` methods in the printers and adding a parenthesize condition.

**Comment by Upabjojr** [comment]
Parenthesize should match the precedence of `Add` non-strictly.

**Comment by Upabjojr** [comment]
Alternatively, modify the constructor of `MatrixElement` in order to distribute the addends over an `Add` object.

**Comment by asmeurer** [comment]
Side issue: why does it print `(-1)*B` instead of `-B`? The MatrixExpression printers maybe ought to be reusing the normal (noncommutative) expression printers more.

**Comment by Upabjojr** [comment]
All printing instances:
https://github.com/sympy/sympy/search?utf8=%E2%9C%93&q=_print_MatrixElement

In the case of `str`:
https://github.com/sympy/sympy/blob/d9c5013222f45d6d9189180e54ba87c69eed8d09/sympy/printing/str.py#L234

**Comment by Alsheh** [comment]
I am working this.

**Comment by Upabjojr** [comment]
I think any expression that is not an atom should be parenthesized.

If possible, use `parenthesize`, examples:
https://github.com/sympy/sympy/search?utf8=%E2%9C%93&q=parenthesize&type=Code

The precedence module:
https://github.com/sympy/sympy/blob/d9c5013222f45d6d9189180e54ba87c69eed8d09/sympy/printing/precedence.py

**Comment by codegeek96** [comment]
I'm new, and would like to solve this as my first bug. Can anyone tell me in which file this issue is?

**Comment by pramod1997** [comment]
I'm new, and would like to solve this . Please help me in getting started.

**Comment by rammanoj** [comment]
Can anyone give more information about this issue?

**Comment by prateeksingh0001** [comment]
Hi, I'm new and I want to start with this issue. 
I've studied the code for the relevant functions and classes but, I'm not able to figure out how to implement the parenthesize function, as in how would we set the level(precedence) for the expression that would be put inside the parenthesize function ?

**Comment by LejlaMetohajrova** [comment]
I made a patch that fixes this issue (#12772). However it is not implemented for rust language yet due to the issue #12773 
Could someone please check that or navigate me how to solve that? Thanks.

**Comment by czgdp1807** [comment]
Current status,
```python
Python 3.6.8 (default, Oct  7 2019, 12:59:55) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy import *
>>> A = MatrixSymbol('A',1,3)
>>> B = MatrixSymbol('B',1,3)
>>> C = MatrixSymbol('C',1,3)
>>> E = A - B
>>> F = C[0,0]
>>> F.subs(C, E)
(A - B)[0, 0]
```
The above result seems correct. So probably the associated open PRs can be closed.
